core/src/main/java/org/elasticsearch/action/bulk/Retry.java

```diff
@@ -0,0 +1,226 @@
+            this.listener = listener;
+            // in contrast to System.currentTimeMillis(), nanoTime() uses a monotonic clock under the hood
+            this.startTimestampNanos = System.nanoTime();
+        }
+
+        @Override
+        public void onResponse(BulkResponse bulkItemResponses) {
+            if (!bulkItemResponses.hasFailures()) {
+                // we're done here, include all responses
+                addResponses(bulkItemResponses, (r -> true));
+                BulkResponse aggregatedResults = getCompleteResponse();
+                notifyListener(aggregatedResults);
+            } else {
+                if (canRetry(bulkItemResponses)) {
+                    addResponses(bulkItemResponses, (r -> !r.isFailed()));
+                    retry(createBulkRequestForRetry(bulkItemResponses));
+                } else {
+                    addResponses(bulkItemResponses, (r -> true));
+                    BulkResponse aggregatedResults = getCompleteResponse();
+                    notifyListener(aggregatedResults);
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            try {
+                listener.onFailure(e);
+            } finally {
+                s.shutdownNow();
+            }
+        }
+
+        private void retry(BulkRequest bulkRequestForRetry) {
+            if (backoff.hasNext()) {
+                s.schedule((() -> this.execute(bulkRequestForRetry)), backoff.next().millis(), TimeUnit.MILLISECONDS);
+            }
+        }
+
+        private BulkRequest createBulkRequestForRetry(BulkResponse bulkItemResponses) {
+            BulkRequest requestToReissue = new BulkRequest();
+            int index = 0;
+            for (BulkItemResponse bulkItemResponse : bulkItemResponses.getItems()) {
+                if (bulkItemResponse.isFailed()) {
+                    // Assumption: ES maintains order between request and response
+                    requestToReissue.add(currentBulkRequest.requests().get(index));
+                }
+                index++;
+            }
+            return requestToReissue;
+        }
+
+        private boolean canRetry(BulkResponse bulkItemResponses) {
+            if (!backoff.hasNext()) {
+                return false;
+            }
+            for (BulkItemResponse bulkItemResponse : bulkItemResponses) {
+                if (bulkItemResponse.isFailed()) {
+                    Throwable cause = bulkItemResponse.getFailure().getCause();
+                    Throwable rootCause = ExceptionsHelper.unwrapCause(cause);
+                    if (!rootCause.getClass().equals(retryOnThrowable)) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private void notifyListener(BulkResponse response) {
+            try {
+                listener.onResponse(response);
+            } finally {
+                s.shutdownNow();
```

Comment:
I'm not sure why, but the usual convention for freeing resources when the request is done is to have a method called `finishHim`. Mortal combat reference? Anyway, the nice thing about this convention is that it gives us a place to look for resources to be freed. But above I mention reusing a thread pool of some sort anyway.